pdns/recursordist/taskqueue.cc

```diff
@@ -0,0 +1,70 @@
+/*
+ * This file is part of PowerDNS or dnsdist.
+ * Copyright -- PowerDNS.COM B.V. and its contributors
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of version 2 of the GNU General Public License as
+ * published by the Free Software Foundation.
+ *
+ * In addition, for the avoidance of any doubt, permission is granted to
+ * link this program with OpenSSL and to (re)distribute the binaries
+ * produced as the result of such linking.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+ */
+
+#include "taskqueue.hh"
+#include "syncres.hh"
+
+extern std::shared_ptr<SyncRes::domainmap_t> g_initialDomainMap;
+
+namespace pdns
+{
+
+bool TaskQueue::empty() const
+{
+  const std::lock_guard<std::mutex> lock(d_mutex);
+  return d_queue.empty();
+}
+
+void TaskQueue::push(const ResolveTask&& task)
+{
+  const std::lock_guard<std::mutex> lock(d_mutex);
+  // Insertion fails if it's already there, no problem since we're already scheduled
+  // and the deadline would remain the same anyway.
+  d_queue.insert(std::move(task));
+}
+
+ResolveTask TaskQueue::pop()
+{
+  std::unique_lock<std::mutex> lock(d_mutex);
+  ResolveTask ret = d_queue.get<SequencedTag>().front();
+  d_queue.get<SequencedTag>().pop_front();
+  return ret;
+}
+
+void TaskQueue::runner()
+{
+  while (!d_queue.empty()) {
+    ResolveTask task = pop();
```

Comment:
There will be a race here if we ever get more than one reader.

Reply:
Yes, agreed. I took a shortcut assuming there is only one reader.